### PR TITLE
Unsigned divmod

### PR DESCRIPTION
**Pull request by jra4**
Add unsigned division primitives (e.g. `%int64#_unsigned_div` and `%int64#_unsigned_mod`).

These are the first unsigned primitives for which small ints operands must be zero extended (rather than signed extended).

Support is added for native and bytecode.

The functions `unsigned_div` and `unsigned_rem` in `Int8`, `Int8_u`, `Int16`, and `Int16_u` have been replaced with the primitives. The other int modules have not been changed as they have many user, and we'd want to be very sure that the primitives match the behavior of the existing functions before swapping them out.

## Testing

All of the `typing_layouts-*/test_*_u.ml` tests have been extended to test unsigned division, along with explicit tests of edge cases.

`division_by_constant.ml` and `divint.ml` have been extended to test the correctness of optimizing known divisors and dividends.

`lib-smallint/test_int{8,16}.ml` tests unsigned division of tagged ints.

Codegen tests have been added.

## Reviewing

Most of the interesting changes are in these files:

  - `to_cmm_primitive.ml`: Does the zero extension for small int operands.
  - `cmm_helpers.ml`:
      - `unsigned_untag_int`: I didn't perform any optimizations like in `untag_int` since it looks like the smart constructor `lsr_const` would handle them anyway.
      - `udivimm_parameters`, `unsigned_div_int`, and `unsigned_mod_int`: Optimize divisions when the operands are known. Known divisor with unknown dividend is the most complicated.
  - `runtime{,4}/ints.c`: Add the unsigned division C functions that are called from bytecode.

Most of the other changes are piping signedness info or adding `unsigned_div` and `unsigned_rem` to flambda int types.

### CODE REVIEW

File: runtime4/ints.c [hunk 1]
```diff
@@ -472,6 +504,15 @@ CAMLexport value caml_copy_int64(int64_t i)
                                                                         \
   CAMLprim int64_t caml_int64_##name##_native
 
+#define CAMLprim_int64_3(name)                                          \
```
Comment (dvulakh) [inline runtime4/ints.c]: I think CAMLprim_uint64_2 could make more sense as a name here

Comment (jra4) [reply runtime4/ints.c]: Yeah, I didn't realize the number represented arity. Updated. Changing the macro name requires a change to `gen_primitives.sh`.

**Review by dvulakh — COMMENTED** [review]
This looks good to me, @mshinwell could you also take a quick look?

File: middle_end/flambda2/to_jsir/to_jsir_primitive.ml [hunk 2]
```diff
@@ -388,8 +388,10 @@ let binary_exn ~env ~res (f : Flambda_primitive.binary_primitive) x y =
       | Add -> "add"
       | Sub -> "sub"
       | Mul -> "mul"
-      | Div -> "div"
-      | Mod -> "mod"
+      | Div Signed -> "div"
+      | Div Unsigned -> "unsigned_div"
```
Comment (julesjacobs) [inline middle_end/flambda2/to_jsir/to_jsir_primitive.ml:392]: (Drive-by AI review -- feel free to ignore & do not lower real review scrutiny)

This appears to lower unsigned div/mod to JSIR externs that the imported JSoO runtime does not provide. A minimal repro using `external unsigned_div : int -> int -> int = "%int_unsigned_div"` and `ocamlj.opt -c -djsir` emits:

    v5 = "%int_unsigned_div"(v1, v2)

The usual provider path for these names is a `//Provides:` / `//Alias:` entry under `external/js_of_ocaml/runtime/js/*.js` (or an internal jsoo primitive registration). I found signed `%int_div` support in `runtime/js/ints.js`, but no `%int_unsigned_div`, `%int_unsigned_mod`, or corresponding `caml_*_unsigned_*` provider/alias. Should these be implemented for JSIR, or should unsigned div/mod be rejected as unsupported there?

File: middle_end/flambda2/parser/fexpr_prim.ml [hunk 3]
```diff
@@ -83,8 +83,8 @@ let binary_int_arith_op =
       ([ "add", Add;
          "sub", Sub;
          "mul", Mul;
-         "div", Div;
-         "mod", Mod;
+         "div", Div Signed;
```
Comment (julesjacobs) [inline middle_end/flambda2/parser/fexpr_prim.ml:86]: (Drive-by AI review -- feel free to ignore & do not lower real review scrutiny)

The fexpr codec still has no spelling for `Div Unsigned` / `Mod Unsigned`. This table only registers `"div" -> Div Signed` and `"mod" -> Mod Signed`, so dumping fexpr for a program using `%int_unsigned_div` crashes in `OfFlambda` with:

    Fatal error: Undefined constructor

I can reproduce this with `ocamlopt.opt -c -dfexpr` on a tiny file containing `external unsigned_div : int -> int -> int = "%int_unsigned_div"`. It looks like fexpr needs distinct unsigned spellings or a signedness parameter here.

File: middle_end/flambda2/to_jsir/to_jsir_primitive.ml (hunk 2 above)
Comment (jra4) [reply middle_end/flambda2/to_jsir/to_jsir_primitive.ml:392]: I don't think we ever got JSIR production-ready, and it has bitrotted over the past few months, so I didn't put too much thought into this. Adding unsigned div to the runtime will be easy anyway if we ever try to resurrect JSIR, though.

File: middle_end/flambda2/parser/fexpr_prim.ml (hunk 3 above)
Comment (jra4) [reply middle_end/flambda2/parser/fexpr_prim.ml:86]: I'll let @mshinwell decide if I need to implement fexpr parser support. I was under the impression that this parser has bitrotted, but I might be wrong

Comment (jra4) [reply middle_end/flambda2/parser/fexpr_prim.ml:86]: I was wrong. `expect_fexpr` is used now, so I updated this.

File: backend/cmm_helpers.ml [hunk 4]
```diff
@@ -4197,6 +4371,14 @@ let mod_int_caml arg1 arg2 dbg =
        (untag_int arg2 dbg) dbg)
     dbg
 
+let unsigned_mod_int_caml arg1 arg2 dbg =
+  tag_int
+    (unsigned_mod_int
```
Comment (ttebbi-janestreet) [inline backend/cmm_helpers.ml:4380]: Since Ocaml ints are 63 bit, I'm pretty sure we could emit better code for them, with just a single multiplication and not the additional fixup of the `a=true` case. Since the details are so complicated, maybe let's just add a CR saying so?

File: middle_end/flambda2/simplify/number_adjuncts.ml [hunk 5]
```diff
@@ -518,8 +524,14 @@ module For_int8s : Int_number_kind = struct
 
     let div x y = of_int (Int.div (to_int x) (to_int y))
 
+    let unsigned_div x y =
+      of_int (Int.div (unsigned_to_int x) (unsigned_to_int y))
+
     let rem x y = of_int (Int.rem (to_int x) (to_int y))
 
+    let rem_unsigned x y =
```
Comment (ttebbi-janestreet) [inline middle_end/flambda2/simplify/number_adjuncts.ml]: nit: Why not `unsigned_rem`?

**Review by ttebbi-janestreet — APPROVED** [review]

File: backend/cmm_helpers.ml (hunk 4 above)
Comment (jra4) [reply backend/cmm_helpers.ml:4380]: I verified this was the case and compared with Clang. A divisor of 7 is such an example, and Clang is able to pick parameters with `a=false` if it knows the top bit is zero. Added a CR-someday.

File: middle_end/flambda2/simplify/number_adjuncts.ml (hunk 5 above)
Comment (jra4) [reply middle_end/flambda2/simplify/number_adjuncts.ml]: Changed it to `unsigned_rem`
